tests/virtctl/usbredir.go

```diff
@@ -0,0 +1,271 @@
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00,
+}
+
+type ctxKeyType string
+
+const connectedKey ctxKeyType = "connected"
+
+var _ = Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component][sig-compute] USB Redirection", decorators.SigCompute, func() {
+	const (
+		enoughMemForSafeBiosEmulation = "32Mi"
+		vmiRunTimeout                 = 90
+	)
+
+	Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component] A VirtualMachineInstance with usbredir support", func() {
+		var vmi *v1.VirtualMachineInstance
+
+		BeforeEach(func() {
+			// A VMI for each test to have fresh stack on server side
+			vmi = libvmi.New(libvmi.WithResourceMemory(enoughMemForSafeBiosEmulation), withClientPassthrough())
+			vmi = libvmops.RunVMIAndExpectLaunch(vmi, vmiRunTimeout)
+		})
+
+		It("Should fail when limit is reached", func() {
+			type session struct {
+				cancel context.CancelFunc
+				err    chan error
+			}
+			var tests []session
+			for i := range v1.UsbClientPassthroughMaxNumberOf + 1 {
+				Eventually(func(g Gomega) bool {
+					cmd := newVirtctlCommand("usbredir",
+						"--namespace", vmi.ObjectMeta.Namespace,
+						"--no-launch", vmi.ObjectMeta.Name)
+					connect := make(chan struct{})
+					ctx := context.WithValue(cmd.Context(), connectedKey, connect)
+					ctx, cancel := context.WithCancel(ctx)
+					cmd.SetContext(ctx)
+
+					err := make(chan error, 1)
+					go runConnectGoroutine(cmd, err)
+
+					if i == v1.UsbClientPassthroughMaxNumberOf {
+						// Last test is meant to fail.
+						tests = append(tests, session{cancel, err})
+						return true
+					}
+
+					// Till the last test, all sockets must be connected
+					select {
+					case <-connect:
+						tests = append(tests, session{cancel, err})
+						return true
+					case err := <-err:
+						Expect(err).To(MatchError(syscall.ECONNRESET))
+					}
+					cancel()
+					return false
+				}, 5*time.Second, 1*time.Second).Should(BeTrue())
+			}
+
+			numOfErrors := 0
+			for i := range v1.UsbClientPassthroughMaxNumberOf + 1 {
+				select {
+				case err := <-tests[i].err:
+					Expect(err).To(MatchError(ContainSubstring("websocket: bad handshake")))
+					numOfErrors++
+					tests[i].cancel()
+				case <-time.After(time.Second):
+					tests[i].cancel()
+				}
+			}
+			Expect(numOfErrors).To(Equal(1), "Only one connection should fail")
+		})
+
+		It("Should work several times", func() {
+			for range 4 * v1.UsbClientPassthroughMaxNumberOf {
+				Eventually(func(g Gomega) bool {
+					cmd := newVirtctlCommand("usbredir",
+						"--namespace", vmi.ObjectMeta.Namespace,
+						"--no-launch", vmi.ObjectMeta.Name)
+					connect := make(chan struct{})
+					ctx := context.WithValue(cmd.Context(), connectedKey, connect)
+					ctx, cancel := context.WithTimeout(ctx, time.Second)
+					cmd.SetContext(ctx)
```

Comment:
Might be worth de-duping this block. It's the same in the test above.

Reply:
I thought about it but one is WithCancel the other WithTimeout and we must cmd.SetContext() for the runConnectGoroutine. Not sure if worth the trouble

Reply:
Let's keep it.